### PR TITLE
Recipe Shortcode Updates

### LINKED ISSUE #4470: [Recipe] shortcode: add additional microformats support (v1 and v2)
#### Steps to reproduce the issue

The current [recipe] shortcode appears to _only_ support the '[hrecipe](http://microformats.org/wiki/hRecipe)' class in microformats1
#### What I expected

I would have expected the additional dozen or so classes in hrecipe to be supported, or better yet, that it would have also supported the newer microformats2 [h-recipe](http://microformats.org/wiki/h-recipe) spec in parallel.

I'm anything but a github expert, but I took a crack at adding the additional classes (for both mf1 and mf2) on a forked branch, which also includes some additional detail:
https://github.com/chrisaldrich/jetpack/pull/1/commits/f914ebe3bd573d1d6d42c5b7752bf26c41448a3b

I've got a copy of this patch running on a site with a recipe at: http://boffosocko.com/2016/07/17/chili-prep-for-dinner-tonight/

### PR DESCRIPTION
**Pull request by ChaosExAnima**
Fixes #4470, #7878, #13660 

#### Changes proposed in this Pull Request:

This updates the recipe shortcode as follows:
* Adds Cooking Time, Prep Time, and Rating attributes to `[recipe]` shortcode.
* Adds `[recipe-nutrition]` shortcode to allow for nutrition listing.
* Adds `[recipe-image]` shortcode to control placement of recipe image.
* Updates image output to allow for WordPress attachment as well as URL.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* p9dueE-14w-p2

#### Testing instructions:

* Create a post with a recipe as per [WP.com article](https://en.support.wordpress.com/recipes/)
* Add `cookingtime`, `preptime`, and `rating` attributes to `recipe` shortcode.
* Add `recipe-nutrition` shortcode.
* Add `recipe-image` shortcode with attachment ID.

#### Proposed changelog entry for your changes:

* Enhances recipe shortcode with additional customization.

### CODE REVIEW

**Comment by ChaosExAnima** [comment]
Messed up a bit in git history so had to rebase, thus all the `not verified`.

**Review by jeherve — COMMENTED** [review]
I tested this a bit and would have some remarks below.

When I test using [this shortcode](https://gist.github.com/jeherve/d46c9a06453362ef8cdee45def57b40e), I get a number of errors in the schema validator here:
https://search.google.com/structured-data/testing-tool/u/0/

image

Do you get such errors as well?

I don't seem to get those errors with the current version of the recipe shortcode ([here is an example](https://search.google.com/structured-data/testing-tool/u/0/#url=https%3A%2F%2Fjeremy.hu%2Fapricot-cake-a-quick-recipe%2F)).

File: modules/shortcodes/recipe.php [hunk 1]
```diff
@@ -313,12 +301,36 @@ private static function recipe_shortcode_html( $atts, $content = '' ) {
...
+	 *
+	 * @param array $atts Array of shortcode attributes.
+	 *
+	 * @return string HTML for recipe notes shortcode.
+	 */
+	public static function recipe_image_shortcode( $atts ) {
+		$atts = shortcode_atts(
+			array(
+				'image' => '', // string.
+				0       => '', // string.
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: That's the positional aspect for a shortcode: `[foo bar] turns into `[ 0 => 'bar' ]`

File: modules/shortcodes/recipe.php [hunk 2]
```diff
@@ -508,12 +567,86 @@ public static function recipe_directions_shortcode( $atts, $content = '' ) {
...
+				$src,
+				'full',
+				false,
+				array(
+					'class'    => 'jetpack-recipe-image u-photo',
+					'itemprop' => 'image',
+				)
+			);
+		}
+		return sprintf(
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: We can test for it being a valid URL, but beyond that I'd think would be overkill.

File: modules/shortcodes/recipe.php [hunk 3]
```diff
@@ -251,6 +218,28 @@ private static function recipe_shortcode_html( $atts, $content = '' ) {
 				);
 			}
 
+			if ( '' !== $atts['rating'] ) {
+				$html .= sprintf(
+					'<li class="jetpack-recipe-rating">
+						<strong>%1$s: </strong>
+						<span itemprop="contentRating">%2$s</span>
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: So I checked it out, and the [Google Structured Testing Tool](https://search.google.com/structured-data/testing-tool) realllyy didn't like `AggregateRating`. And to be fair it's not an aggregate, it's the rating in the content- we only specify one source, not multiple.

File: modules/shortcodes/recipe.php [hunk 4]
```diff
@@ -251,6 +218,28 @@ private static function recipe_shortcode_html( $atts, $content = '' ) {
 				);
 			}
 
+			if ( '' !== $atts['rating'] ) {
+				$html .= sprintf(
+					'<li class="jetpack-recipe-rating">
+						<strong>%1$s: </strong>
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: I agree, though I'd definitely want to ensure we don't go too deeply down the rabbit hole here.

File: modules/shortcodes/recipe.php [hunk 5]
```diff
@@ -31,67 +31,36 @@ class Jetpack_Recipes {
...
-	 * @param string $tag      New HTML tag to add to the array of allowed HTML.
-	 * @param array  $rules    Array of allowed attributes for that HTML tag.
+	 * @return array Array to be used by KSES.
 	 */
-	private function add_kses_rule( $all_tags, $tag, $rules ) {
-
-		// If the tag doesn't already exist, add it.
-		if ( ! isset( $all_tags[ $tag ] ) ) {
-			$all_tags[ $tag ] = array();
+	private static function kses_tags() {
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: Done in bf449dd

File: modules/shortcodes/recipe.php [hunk 6]
```diff
@@ -183,6 +154,9 @@ public static function recipe_shortcode( $atts, $content = '' ) {
 				'sourceurl'   => '', // string.
 				'image'       => '', // string.
 				'description' => '', // string.
+				'cooktime'    => '', // string.
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: Done in 5da8961

File: modules/shortcodes/recipe.php [hunk 7]
```diff
@@ -251,6 +218,28 @@ private static function recipe_shortcode_html( $atts, $content = '' ) {
...
+					'<li class="jetpack-recipe-rating">
+						<strong>%1$s: </strong>
+						<span itemprop="contentRating">%2$s</span>
+					</li>',
+					esc_html_x( 'Rating', 'recipe', 'jetpack' ),
+					esc_html( $atts['rating'] )
+				);
+			}
+
+			if ( '' !== $atts['rating'] ) {
```
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: Done in d3881c0

File: modules/shortcodes/recipe.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChaosExAnima) [reply modules/shortcodes/recipe.php]: Done in 9c9e1a8. Note this solution will fail for relative paths from the post, but I think that's acceptable. Do you concur?

**Comment by ChaosExAnima** [comment]
@jeherve - I did some testing and fixes for the microformats markup. A few notes:

- `preptime` and `cooktime` are indicated in the schema tester, but they appear to be Google exclusive extensions and not in the [original spec](http://microformats.org/wiki/hrecipe). I got them as warnings but they aren't parseable.
- `description` isn't working as expected and also a Google extension.

File: modules/shortcodes/recipe.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeherve) [reply modules/shortcodes/recipe.php]: Yes, I think that's fine. I think WordPress site owners are more used to absolute URLs overall.

**Review by jeherve — APPROVED** [review]
This looks good to me. 👍

**Pull request merged**